README.md

<p align="left">
    <b><a href="https://github.com/katzer/cordova-plugin-printer/blob/example/README.md">SAMPLE APP</a> :point_right:</b>
</p>

# Cordova Print Plugin <br> [![npm version](https://badge.fury.io/js/cordova-plugin-printer.svg)](http://badge.fury.io/js/cordova-plugin-printer) [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0) [![PayPayl donate button](https://img.shields.io/badge/paypal-donate-yellow.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=L3HKQCD9UA35A "Donate once-off to this project using Paypal")

Plugin for [Cordova][cordova] to print documents, photos, HTML and plain text from iOS, Android and Windows Universal apps.

```js
cordova.plugins.printer.print('<b>Hello Cordova!</b>');
```

<img width="280px" align="right" src="https://github.com/katzer/cordova-plugin-printer/blob/example/images/print.png">

### Supported Printer Interfaces

- Apple AirPrint
- Android Print
- Windows Print

### Supported Content

- HTML
- Text
- Base64
- Images
- PDF

### Supported Platforms

- Android 4.4+
- iOS 10+
- Windows 10 UWP
- Browser

## Basics

The plugin creates the object `cordova.plugins.printer` and is accessible after the *deviceready* event has been fired.

```js
document.addEventListener('deviceready', function () {
    // cordova.plugins.printer is now available
}, false);
```

Prints the contents of the web view:

```javascript
cordova.plugins.printer.print();
```

Plain text:

```javascript
cordova.plugins.printer.print("Hello\nWorld!");
```

HTML & CSS:

```javascript
cordova.plugins.printer.print('<h1>Hello World!</h1>');
```

Images, PDF and other documents:

```javascript
cordova.plugins.printer.print('file://img/logo.png');
```

Base64 encoded content:

```javascript
cordova.plugins.printer.print('base64://...');
```

__Note:__ On the browser platform the plugin only supports to print the contents of the web view.

## Formatting

It's possible to pass format options to the print method that overrides the defaults:

```javascript
cordova.plugins.printer.print(content, options, callback);
```

The defaults are defined as follows:

```javascript
cordova.plugins.printer.setDefaults({ monochrome: true });
```

The list of possible options depend on the platform, the content type and the capabilities of the printer.

| Name | Description | Type | Platform |
|:---- |:----------- |:----:| --------:|
| name | The name of the print job and of the document. | String | all |
| copies | The number of copies for the print task. | Number | iOS<br>Windows |
| pageCount | Limits the pages to print even the document contains more.<br>To skip the last n pages you can assign a negative value on iOS. | Number | iOS<br>Android |
| duplex | Either double-sided on short site (duplex:'short'), double-sided on long site (duplex:'long') or single-sided (duplex:'none'). | String | all |
| orientation | The orientation of the printed content, `portrait` or `landscape`. | String | all |
| monochrome | If your application only prints black text, setting this property to _true_ can result in better performance in many cases. | Boolean | all |
| photo | Set to _true_ to change the media type to photography for higher quality. | Boolean | iOS<br>Windows |
| autoFit | Set to _false_ to disable downscaling the image to fit into the content aread. | Boolean | Android |
| printer | The network URL to the printer. | String | iOS |
| maxHeight<br>maxWidth | Defines the maximum size of the content area. | Unit | iOS |
| margin | Set to _false_ to avoid margins. | Boolean | all |
| margin.top<br>margin.left<br>margin.right<br>margin.bottom | The margins for each printed page. Each printer might have its own minimum margins depends on media type and paper format. | Unit | iOS |
| ui.hideNumberOfCopies | Set to _true_ to hide the control for the number of copies. | Boolean | iOS |
| ui.hidePaperFormat | Set to _true_ to hide the control for the paper format. | Boolean | iOS |
| ui.top<br>ui.left | The position of the printer picker. | Number | iPad |
| ui.height<br>ui.width | The size of the printer picker. | Number | iPad |
| paper.width<br>paper.height | The dimensions of the paper – iOS will will try to choose a format which fits bests. | Unit | iOS |
| paper.name | The name of the format like `IsoA4` or `Roll22Inch`.<br>https://docs.microsoft.com/en-us/uwp/api/windows.graphics.printing.printmediasize | String | Windows |
| paper.length | On roll-fed printers you can decide when the printer cuts the paper. | Unit | iOS |
| font.name | The name of the font family | String | iOS |
| font.size | The size of the font | Number | iOS<br>Android |
| font.italic<br>font.bold | Set to _true_ to enable these font traits. | Boolean | iOS |
| font.align | Possible alignments are `left`, `right`, `center` and `justified`. | String | iOS |
| font.color | The color of the font in hexa-decimal RGB format - `"FF0000"` means red. | String | iOS |
| header.height<br>footer.height | The height of the header or footer on each page. | Unit | iOS |
| header.labels<br>footer.labels | An array of labels to display. Only use if there are more then one. | Array | iOS |
| header.label.text<br>footer.label.text | The plain text to display. Use `%ld` to indicate where to insert the page index.<br>For example `"Page %ld"` would result into `"Page 1"`, `"Page 2"`, ... | String | iOS |
| header.label.top<br>header.label.right<br>header.label.left<br>header.label.bottom<br>footer.label.* | The relative position where to place the label within the footer or header area. | Unit | iOS |
| header.label.font<br>footer.label.font | The font attributes for the label. | Object | iOS |
| header.label.showPageIndex<br>footer.label.showPageIndex | Set to _true_ if you want to display the page index.<br> | Boolean | iOS |

The `Unit` type can be either a (float) number or a string with a special suffix.

- Supported unit suffixes are `in` for inches, `mm` for millimeters, `cm` for centimeters and `pt` for points
- `"2in"` are two inches whereas `2.0` or `"2.0pt"` are identical for two points
- One inch are 72.0 points

## Direct Print

For iOS its possible to send the content directly to the printer without any dialog. Todo so pass the network URL as an option:

```javascript
cordova.plugins.printer.print(content, { printer: 'ipp://...' });
```

To let the user pick an available printer:

```javascript
cordova.plugins.printer.pick(function (url) {});
```

It's possible to specify the position of the picker:

```javascript
cordova.plugins.printer.pick({ top: 40, left: 30 }, callback);
```

__Note:__ By passing an invalid URL, the application will throw an `Unable to connect to (null)` exception and possibly crash.

## Printable Document Types

The list of supported document types differ between mobile platforms. As of writing, Windows UWP only supports HTML and plain text.

To get a list of all printable document types:

```javascript
cordova.plugins.printer.getPrintableTypes(callback);
```

To check if printing is supported in general:

```javascript
cordova.plugins.printer.canPrintItem(callback);
```

Or in particular:

```javascript
cordova.plugins.printer.canPrintItem('file://css/index.css', callback);
```

## Sample

```js
var options = {
    font: {
        size: 22,
        italic: true,
        align: 'center'
    },
    header: {
        height: '6cm',
        label: {
            text: "\n\nDie Freuden",
            font: {
                bold: true,
                size: 37,
                align: 'center'
            }
        }
    },
    footer: {
        height: '4cm',
        label: {
            text: 'Johann Wolfgang von Goethe, 1749-1832, deutscher Dichter, Naturforscher',
            font: { align: 'center' }
        }
    }
};

cordova.plugins.printer.print("Es flattert um die Quelle\nDie wechselnde Libelle,...", options);
```

The result will look like this for iOS:

![ttt](https://github.com/katzer/cordova-plugin-printer/blob/example/images/sample.png)

## Installation

Execute from the projects root folder:

    $ cordova plugin add cordova-plugin-printer

Or install a specific version:

    $ cordova plugin add cordova-plugin-printer@VERSION

Or install the latest head version:

    $ cordova plugin add https://github.com/katzer/cordova-plugin-printer.git

Or install from local source:

    $ cordova plugin add <path> --nofetch --nosave

Then execute:

    cordova build

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License

This software is released under the [Apache 2.0 License][apache2_license].

Made with :yum: from Leipzig

© 2013 [appPlant GmbH][appplant]


[cordova]: https://cordova.apache.org
[apache2_license]: http://opensource.org/licenses/Apache-2.0
[appplant]: www.appplant.de


CHANGELOG.md
## ChangeLog

#### Version 0.8.0 (01.03.2019)
- Most parts of the code has been rewritten
- Added support for Android 8+
- Added support for iOS 10+
- Added support for the browser platform
- Added support for header and foooter
- Added support to print PDF files, images, Base64 encoded images
- And many more, see the new README

#### Version 0.7.3 (19.12.2016)
- Fixed incompatibility with Android KitKat (4.4)

#### Version 0.7.2 (03.08.2016)
- [__change__:] Changed plugin ID to `cordova-plugin-printer`
- [__change__:] Plugin requires Android KitKat or newer
- [__change__:] `isAvailable` returns false if no enabled service can be found (Android)
- [feature:] New `pick` interface to pick a printer for future usage
- [feature:] Support for Windows10 platform (Thanks to #cristi-badila)
- [enhancement:] `isAvailable` returns count of available services (Android)
- [enhancement:] `print` returns bool value to indicate the result
- [enhancement:] Added missing `duplex` support (Android)
- [__change__:] `duplex` requires a string (`none`, `long` or `short`)
- [enhancement:] Support for `border`, `hidePageRange`, `hideNumberOfCopies` and `hidePaperFormat` (iOS specific)

#### Version 0.7.1 (23.04.2015)
- [bugfix:] `isAvailable` does not block the main thread anymore.
- [bugfix:] iPad+iOS8 incompatibility (Thanks to __zmagyar__)
- [enhancement:] Print-View positioning on iPad
- [enhancement:] Send direct to printer when printerId: is specified.

#### Version 0.7.0 (12.09.2014)
- Android Printing Framework support
- [__change__:] Renamed `isServiceAvailable` to `isAvailable`
- [enhancement:] New print options like `name`, `landscape` or `duplex`
- [enhancement:] Ability to print remote content via URI
- [enhancement:] Callback support

www/printer.js
/*
 Copyright 2013 Sebastián Katzer

 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 */

var exec  = require('cordova/exec'),
    ua    = navigator.userAgent.toLowerCase(),
    isIOS = ua.indexOf('ipad') > -1 || ua.indexOf('iphone') > -1;

// Defaults
exports._defaults = {
    // name:       'unknown',

    // duplex: 'none',
    // orientation: 'landscape',
    // monochrome: false,
    // photo: false,

    // copies: 1,
    // pageCount: 1,

    // maxHeight: '10cm',
    // maxWidth: '10cm',

    // font: {
    //     name: 'Helvetica',
    //     align: 'left',
    //     italic: false,
    //     bold: false,
    //     color: '#FF0000'
    // },

    // margin: {
    //     top: 0,
    //     left: 0,
    //     bottom: 0,
    //     right: 0
    // },

    // ui: {
    //     hideNumberOfCopies: false,
    //     hidePaperFormat:    false,
    //     top: 30,
    //     left: 40,
    //     height: 0,
    //     width: 0
    // },

    // paper: {
    //     height: 0,
    //     width: 0,
    //     length: 0,
    //     name: 'A4'
    // },

    // header: {
    //     height: '1cm',

    //     labels: [{
    //         text: 'Awesome Printer Plug-in',
    //         font: {
    //             align: 'center',
    //             italic: true,
    //             color: '#FF0000'
    //         }
    //     },{
    //         showPageIndex: true,
    //         font: {
    //             align: 'right',
    //             bold: true
    //         }
    //     }]
    // },

    // footer: {
    //     height: '3mm',

    //     label: {
    //         text: 'Copyright (c) 2013-2019 Sebastián Katzer',
    //         font: { size: 9 },
    //         top: '1.5mm',
    //         right: '5mm'
    //     }
    // }
};

/**
 * Test if the printer service is able to print
 * the ressource specified by URL.
 *
 * @param [ String ]   uri      A file URI.
 * @param [ Function ] callback The callback function.
 * @param [ Object ]   scope    The scope for the function.
 *
 * @return [ Void ]
 */
exports.canPrintItem = function (uri, callback, scope)
{
    if (typeof uri == 'function')
    {
        scope    = callback;
        callback = uri;
        uri      = null;
    }

    var fn = this._createCallbackFn(callback, scope);

    exec(fn, null, 'Printer', 'check', [uri]);
};

/**
 * Returns a list of all printable document types.
 *
 * @param [ Function ] callback The callback function.
 * @param [ Object ]   scope    The scope for the function.
 *
 * @return [ Void ]
 */
exports.getPrintableTypes = function (callback, scope)
{
    var fn = this._createCallbackFn(callback, scope);

    exec(fn, null, 'Printer', 'types', []);
};

/**
 * Displays system interface for selecting a printer.
 *
 * @param [ Object ]   options  Optional ui options.
 * @param [ Function ] callback The callback function.
 * @param [ Object ]   scope    The scope for the function.
 *
 * @return [ Void ]
 */
exports.pick = function (options, callback, scope)
{
    if (typeof options == 'function')
    {
        scope    = callback;
        callback = options;
        options  = {};
    }

    var fn     = this._createCallbackFn(callback, scope),
        params = this._mergeWithDefaults({ ui: options });

    if (isIOS)
    {
        exec(fn, null, 'Printer', 'pick', [params]);
    }
    else if (fn)
    {
        fn(null);
    }
};

/**
 * Sends the content to the printer.
 *
 * @param [ String ]   content  The plain/html text or a file URI.
 * @param [ Object ]   options  Options for the print job.
 * @param [ Function ] callback The callback function.
 * @param [ Object ]   scope    The scope for the function.
 */
exports.print = function (content, options, callback, scope)
{
    if (typeof content == 'function')
    {
        scope    = options;
        callback = content;
        options  = {};
        content  = null;
    }

    if (typeof options == 'function')
    {
        scope    = callback;
        callback = options;
        options  = typeof content != 'string' ? content : {};
        content  = typeof content == 'string' ? content : null;
    }

    var fn     = this._createCallbackFn(callback, scope),
        params = this._mergeWithDefaults(options || {});

    exec(fn, null, 'Printer', 'print', [content || '', params]);
};

/**
 * The (platform specific) default settings.
 *
 * @return [ Object ]
 */
exports.getDefaults = function ()
{
    var map = Object.assign({}, this._defaults);

    for (var key in map)
    {
        if (Array.isArray(map[key]))
        {
            map[key] = Array.from(map[key]);
        }
        else if (Object.prototype.isPrototypeOf(map[key]))
        {
            map[key] = Object.assign({}, map[key]);
        }
    }

    return map;
};

/**
 * Overwrite default settings.
 *
 * @param [ Object ] newDefaults New default values.
 *
 * @return [ Void ]
 */
exports.setDefaults = function (newDefaults)
{
    Object.assign(this._defaults, newDefaults);
};

/**
 * Merge custom properties with the default values.
 *
 * @param [ Object ] options Set of custom values.
 *
 * @retrun [ Object ]
 */
exports._mergeWithDefaults = function (options)
{
    var defaults = this.getDefaults();

    if (options.duplex && typeof options.duplex == 'boolean')
    {
        options.duplex = options.duplex ? 'long' : 'none';
    }

    Object.assign(defaults, options);

    for (var key in defaults)
    {
        if (defaults[key] !== null)
        {
            options[key] = defaults[key];
        }
        else
        {
            delete options[key];
        }
    }

    return options;
};

/**
 * @private
 *
 * Creates a callback, which will be executed
 * within a specific scope.
 *
 * @param [ Function ] callback The callback function.
 * @param [ Object ]   scope    The scope for the function.
 *
 * @return [ Function ] The new callback function
 */
exports._createCallbackFn = function (callback, scope)
{
    if (typeof callback !== 'function')
        return;

    return function () {
        callback.apply(scope || this, arguments);
    };
};

// Polyfill for Object.assign
if (typeof Object.assign != 'function') {
  Object.assign = function(target) {
    'use strict';
    if (target == null) {
      throw new TypeError('Cannot convert undefined or null to object');
    }

    target = Object(target);
    for (var index = 1; index < arguments.length; index++) {
      var source = arguments[index];
      if (source != null) {
        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }
    }
    return target;
  };
}

// Polyfill for Array.from
// Production steps of ECMA-262, Edition 6, 22.1.2.1
// Reference: https://people.mozilla.org/~jorendorff/es6-draft.html#sec-array.from
if (!Array.from) {
  Array.from = (function () {
    var toStr = Object.prototype.toString;
    var isCallable = function (fn) {
      return typeof fn === 'function' || toStr.call(fn) === '[object Function]';
    };
    var toInteger = function (value) {
      var number = Number(value);
      if (isNaN(number)) { return 0; }
      if (number === 0 || !isFinite(number)) { return number; }
      return (number > 0 ? 1 : -1) * Math.floor(Math.abs(number));
    };
    var maxSafeInteger = Math.pow(2, 53) - 1;
    var toLength = function (value) {
      var len = toInteger(value);
      return Math.min(Math.max(len, 0), maxSafeInteger);
    };

    // The length property of the from method is 1.
    return function from(arrayLike/*, mapFn, thisArg */) {
      // 1. Let C be the this value.
      var C = this;

      // 2. Let items be ToObject(arrayLike).
      var items = Object(arrayLike);

      // 3. ReturnIfAbrupt(items).
      if (arrayLike == null) {
        throw new TypeError("Array.from requires an array-like object - not null or undefined");
      }

      // 4. If mapfn is undefined, then let mapping be false.
      var mapFn = arguments.length > 1 ? arguments[1] : void undefined;
      var T;
      if (typeof mapFn !== 'undefined') {
        // 5. else
        // 5. a If IsCallable(mapfn) is false, throw a TypeError exception.
        if (!isCallable(mapFn)) {
          throw new TypeError('Array.from: when provided, the second argument must be a function');
        }

        // 5. b. If thisArg was supplied, let T be thisArg; else let T be undefined.
        if (arguments.length > 2) {
          T = arguments[2];
        }
      }

      // 10. Let lenValue be Get(items, "length").
      // 11. Let len be ToLength(lenValue).
      var len = toLength(items.length);

      // 13. If IsConstructor(C) is true, then
      // 13. a. Let A be the result of calling the [[Construct]] internal method of C with an argument list containing the single item len.
      // 14. a. Else, Let A be ArrayCreate(len).
      var A = isCallable(C) ? Object(new C(len)) : new Array(len);

      // 16. Let k be 0.
      var k = 0;
      // 17. Repeat, while k < len… (also steps a - h)
      var kValue;
      while (k < len) {
        kValue = items[k];
        if (mapFn) {
          A[k] = typeof T === 'undefined' ? mapFn(kValue, k) : mapFn.call(T, kValue, k);
        } else {
          A[k] = kValue;
        }
        k += 1;
      }
      // 18. Let putStatus be Put(A, "length", len, true).
      A.length = len;
      // 20. Return A.
      return A;
    };
  }());
}
